### ISSUE TITLE
Display is corrupted when `temporarily_moved_cursor` fails to restore cursor after displaying an image with `kitty`

### ISSUE THREAD

**Issue by rent-yr-chemicals** [ranger/ranger#2732]
#### Runtime Environment

- Operating system and version: Kubuntu 22.04.1 LTS x86_64
- Terminal emulator and version: kitty 0.26.3
- Python version: Python 3.10.6
- Ranger version/commit: ranger-master v1.9.3-529-g2bc4efdf
- Locale: en_US.UTF-8

#### Current Behavior

When `kitty` is used to natively preview images, the cursor is moved into position with the `temporarily_moved_cursor` context manager; in theory, this put the cursor where `kitty` needs it to be, displays the image, then returns the cursor to its original position. In practice, this either doesn't happen or inconsistently happens(?), resulting in a corrupted display in certain situations; this occurs during the  _next_ after the image was displayed, when `curses` assumes while updating the display that the cursor is still where it left it prior to `temporarily_moved_cursor` being called.

This isn't noticeable under usual circumstances, as the cursor itself is usually hidden, and actions that would trigger a redraw also frequently cause the image to be cleared. See "Steps to Reproduce" for an example situation where this does occur.

#### Expected Behavior

The cursor is restored to its original position after `temporarily_moved_cursor` returns. The display is uncorrupted.

#### Steps to reproduce

1. [optional, makes the problem clearer] Set `show_cursor` to `True`
2. Set `preview_images True`, set `preview_images_method kitty`
3. Enter a directory containing a single image file
4. The image is displayed, as expected; the cursor is now adjacent to the lower right corner of the image
5. Mark the file with the spacebar; since it's the only file in the directory, it's still selected, and the image is still displayed (as expected)
6. Rather than the line being updated reflecting that the file is marked, the new line is displayed starting from the cursor's current (wrong) location to the bottom right of the image.
7. The display is corrected when we navigate away from the file, reload the directory, or redraw the entire browser (e.g. by setting `fm.ui.browser.need_redraw = True`)

Note that results are more unpredictable when changes occur on more than one line of the main browser column (e.g. by marking/unmarking multiple files), though corruption of some kind is generally present; I suspect the precise form is being determined by the internals of `ncurses`.

#### Possible Solutions

In `kitty`'s case, it seems like image display capabilities have been enhanced substantially since ranger's `KittyImageDisplayer` class was last updated. It now appears possible to control image positioning natively via the `kitty`'s graphics protocol itself, which is almost certainly the "correct" solution. I may take a stab at updating ranger's support for `kitty` graphics; if so (emphasis on "if"), I'll be sure to open a pull request.

I have not tested any other terminals whose image display methods make use of `temporarily_moved_cursor`, so I can't speak to whether a similar issue occurs there, but it's likely worth looking into.

**Comment by rent-yr-chemicals** [comment]
Update: I misunderstood `kitty`'s graphics protocol specification; positioning is still required to be relative to the cursor, so the fix will be a pain after all. Working on it.

**Comment by kovidgoyal** [comment]
Use C=1 in the kitty graphics protocol, then the cursor will not be touched by displaying the image. And you can use the save and restore cursor escape codes to make sure the cursor is back to where it was before robustly.

**Comment by rent-yr-chemicals** [comment]
Yep, that's currently what I'm trying. Issue is the restore cursor escape code doesn't take effect immediately (unsure if kitty or ranger is to blame); current fix is to both use C=1 _and_ move the virtual cursor used by curses to the same position, so it knows where to find it even if asked to redraw before the escape code takes effect, but it's a clunky bandaid fix that probably needs improvement

**Comment by kovidgoyal** [comment]
kitty most definitely restores the cursor immediately after seeing the
restore cursor escape code.

**Comment by toonn** [comment]
Thank you for the thorough report with a reproduction!

**Comment by hut** [comment]
> Yep, that's currently what I'm trying. Issue is the restore cursor escape code doesn't take effect immediately (unsure if kitty or ranger is to blame); current fix is to both use C=1 _and_ move the virtual cursor used by curses to the same position, so it knows where to find it even if asked to redraw before the escape code takes effect, but it's a clunky bandaid fix that probably needs improvement

@rent-yr-chemicals you speak of a "current fix" - do you have some code/patch to share?
